### PR TITLE
Improve AuthorizerAdaptorEntrypoint docs

### PR DESCRIPTION
**Pull request by nventuro**
# Description

This goes into much more detail as to what the issue is and how the fix works in all components involved, so that a reader that stumbles on any of them will get sufficient context to understand the full picture.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [x] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptor.sol [hunk 1]
```diff
@@ -92,6 +101,31 @@ contract AuthorizerAdaptor is IAuthorizerAdaptor, ReentrancyGuard {
...
+        // We unconditionally read memory slot 100, which is where the first four bytes of `data` will reside (i.e. the
+        // function selector) given a standard packed ABI encoding. Both the Solidity compiler and clients such as
+        // ethers.js will do the ABI encoding in such a way that the selector is actually on slot 100, since this is the
+        // way that minimizes gas costs, but it is *not* the only valid way to ABI encode.
+        // In particular, it is possible to choose a larger offset and place `data` much further away in calldata. Under
+        // those conditions, slot 100 will *not* contain the selector, but it can instead be any arbitrary value. This
+        // means that the AuthorizerAdaptor can be made to check for the permission of any arbitrary selector,
+        // regardless of the action encoded in `data`.
+        //
+        // In other words, an account that has permission to execute *any* action via de Adaptor can actually execute
```
Comment (EndymionJkb) [inline pkg/liquidity-mining/contracts/admin/AuthorizerAdaptor.sol:115]:
```suggestion
        // In other words, an account that has permission to execute *any* action via the Adaptor can actually execute
```

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptor.sol [hunk 2]
```diff
@@ -92,6 +101,31 @@ contract AuthorizerAdaptor is IAuthorizerAdaptor, ReentrancyGuard {
...
+        // *all* of them: there's no permission granularity.
+        // Note that actually performing this exploit requires the ability to manually craft calldata: as such,
+        // Solidity contracts that call into the Adaptor and create the call via the `abi.encode` function are safe to
+        // use since they will always use the standard encoding.
+        //
+        // To work around this issue, the `TimelockAuthorizer` contract contains a special condition that will check
+        // when it is being called by the `AuthrozierAdaptor`, and behave differently when that happens. See the
```
Comment (EndymionJkb) [inline pkg/liquidity-mining/contracts/admin/AuthorizerAdaptor.sol:122]:
```suggestion
        // when it is being called by the `AuthorizerAdaptor`, and behave differently when that happens. See the
```

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol [hunk 3]
```diff
@@ -71,39 +60,32 @@ contract AuthorizerAdaptorEntrypoint is IAuthorizerAdaptorEntrypoint {
...
 
-        // This call to `canPerform` will validate the actual action ID and sender in the authorizer.
         _require(canPerform(getActionId(selector), msg.sender, target), Errors.SENDER_NOT_ALLOWED);
 
-        // Contracts using the adaptor expect it to be the caller of the actions to perform, so we forward
-        // the call to `performAction` to the adaptor instead of performing it directly.
+        // The `AuthorizerAdaptor` will not check for permissions: it is special-cased in the `TimelockAuthorizer` so
+        // that all calls to it that are not made from this entrypoint fail, while those that originate in the
+        // entrypoint succeed. This works as we have just checked that the caller has permission to perform the action
+        // encoded by `data`. See `TimelockAuthorizer.canPeform` for more details.
```
Comment (EndymionJkb) [inline pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol:88]:
```suggestion
        // encoded by `data`. See `TimelockAuthorizer.canPerform` for more details.
```

File: pkg/vault/contracts/authorizer/TimelockAuthorizer.sol [hunk 4]
```diff
@@ -366,14 +366,16 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
...
             //
-            // To remedy this we force all calls to the authorizer adaptor be through a singleton entrypoint contract
-            // This contract correctly checks whether `account` can perform `actionId` on `where` and forwards the call
-            // onto the authorizer adaptor to execute.
-            // The authorizer must then reject calls to the authorizer adaptor which aren't made through the entrypoint.
+            // To remedy this we force all calls to the `AuthorizerAdaptor` to be made through a singleton entrypoint
+            // contract, called the `AuthorizerAdaptorEntrypoint`. This contract correctly checks whether `account` can
+            // perform `actionId` on `where`, and then forwards the call onto the `AuthorizerAdaptor` to execute.
+            //
+            // The authorizer then reject calls to the `AuthorizerAdaptor` which aren't made through the entrypoint, and
```
Comment (EndymionJkb) [inline pkg/vault/contracts/authorizer/TimelockAuthorizer.sol:376]:
```suggestion
            // The authorizer then rejects calls to the `AuthorizerAdaptor` which aren't made through the entrypoint, and
```

**Review by EndymionJkb — COMMENTED** [review]
Technically the new error message could be separate, but it is related. We have other tests for invalid data that check for "Transaction reverted without a reason"; could call it INVALID_DATA and use it for those cases too. Or say INVALID_CALLDATA / MALFORMED_CALLDATA to differentiate between explicit calldata and general encoded data.

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol [hunk 5]
```diff
@@ -71,39 +60,32 @@ contract AuthorizerAdaptorEntrypoint is IAuthorizerAdaptorEntrypoint {
...
     }
 
-    /**
-     * @notice Performs an arbitrary function call on a target contract, provided the caller is authorized to do so.
-     * @param target - Address of the contract to be called
-     * @param data - Calldata to be sent to the target contract. It should be at least 4 bytes long (i.e. the length of
-     * the selector corresponding to the function to be called)
-     * @return The bytes encoded return value from the performed function call
-     */
     function performAction(address target, bytes calldata data) external payable override returns (bytes memory) {
```
Comment (jubeira) [inline pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol:74]: shouldn't we keep a `@notice` here anyways?

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol (hunk 3 above)
Comment (jubeira) [inline pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol:88]: +1.
Shouldn't we clarify why we are not calling the target contract directly at this point? (that was the intent of the deleted comment)

**Review by jubeira — APPROVED** [review]
Looks good! Just a few minor comments / questions.

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol (hunk 3 above)
Comment (nventuro) [reply pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol:88]: Hm no, I don't think so. That's a different topic (_why_ there's an adaptor in the first place).

File: pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol (hunk 5 above)
Comment (nventuro) [reply pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol:74]: This already exists in the interface.

**Pull request merged**